### PR TITLE
feat(Drawer): add Drawer component

### PR DESCRIPTION
**Pull request by saurabhdaware**
## Documentation Link

https://drawer-docs-tests--61c19ee8d3d282003ac1d81c.chromatic.com/?path=/story/components-drawer

## Description

PR Stack
- Implementation (This PR)
- Docs, Tests #2015 

## Component Checklist

- [ ] Update Component Status Page
- [x] Perform Manual Testing in Other Browsers
- [x] Add KitchenSink Story
- [x] Add Interaction Tests (if applicable)
- [ ] Add changeset

### CODE REVIEW

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts [hunk 1]
```diff
@@ -195,8 +195,8 @@ const buttonIconOnlyPadding: Record<
   xsmall: {
     top: 0,
     bottom: 0,
-    left: 4, // should be `6px` as per design but we're making it `8px` since `6px` is not available as a spacing token
-    right: 4, // should be `6px` as per design but we're making it `8px` since `6px` is not available as a spacing token
+    left: 3, // should be `6px` as per design but we're making it `8px` since `6px` is not available as a spacing token
+    right: 3, // should be `6px` as per design but we're making it `8px` since `6px` is not available as a spacing token
```
Comment (anuraghazra) [inline packages/blade/src/components/Button/BaseButton/buttonTokens.ts:199]: Was it a bug before?

Comment (saurabhdaware) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:199]: yes, The comment says it should be 8px but spacing.4 refers to 12px. It should be 8px only

File: packages/blade/src/components/Drawer/Drawer.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  tags: ['autodocs'],
+  argTypes: {
+    ...getStyledPropsArgTypes(),
+    icon: {
+      name: 'icon',
+      type: 'select' as 'string',
+      options: Object.keys(iconMap),
+      mapping: iconMap,
+    },
+    _isVirtuallyFocussed: {
```
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.stories.tsx:24]: Ignore it. I removed this in my documentation PR. I copy pasted Tag.stories.tsx so it came from that 😆

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts (hunk 1 above)
Comment (anuraghazra) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:199]: Wait this are in pixels or in our spacing token scale? 

does 3 refers to spacing.3 or 3px?

Comment (saurabhdaware) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:199]: It refers to spacing.3

Comment (anuraghazra) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:199]: hmm okay, 

nit: maybe we should instead maintain `spacing.3` instead of 3 as a number.

File: packages/blade/src/components/Drawer/Drawer.web.tsx [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                position="fixed"
+                top="spacing.0"
+                left="100%"
+                backgroundColor="popup.background.subtle"
+                elevation="highRaised"
+                height="100%"
+                ref={refs.setFloating}
+                onKeyDown={(event) => {
+                  if (event?.key === 'Escape' || event?.code === 'Escape') {
+                    onDismiss();
```
Comment (anuraghazra) [inline packages/blade/src/components/Drawer/Drawer.web.tsx:151]: The floating ui dismiss hook didn't work?

File: packages/blade/src/components/Drawer/Drawer.web.tsx [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+ * ```
+ *
+ *  ---
+ *
+ * Checkout {@link https://blade.razorpay.com/?path=/docs/components-drawer Drawer Documentation}
+ * 
+ * 
+ */
+const Drawer = assignWithoutSideEffects(_Drawer, {
+  componentId: drawerComponentIds.Drawer,
```
Comment (anuraghazra) [inline packages/blade/src/components/Drawer/Drawer.web.tsx:203]: Put this id inside `MetaConstants`

Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:203]: MetaConstants is for component attributes no? I have added another entry there. This is for componentIds of individual components

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 3 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:151]: nope. It only works when reference is also handled internally but in our case consumers handle reference

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts (hunk 1 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:199]: yes. I have to fix that size thing for all icon buttons so I'll be creating separate PR for that

File: packages/blade/src/components/Drawer/Drawer.web.tsx [hunk 5]
```diff
@@ -133,7 +130,12 @@ const _Drawer = ({
               ) : null}
               <AnimatedDrawerContainer
                 className={isVisible ? SHOW_DRAWER : ''}
-                width={{ base: '90%', s: '375px', m: '420px' }}
+                isFirstDrawerInStack={isFirstDrawerInStack}
+                width={{
+                  base: stackingLevel > 1 ? 'calc(90% - 16px)' : '90%',
```
Comment (anuraghazra) [inline packages/blade/src/components/Drawer/Drawer.web.tsx:135]: Is this for mobile? 

Let's properly test multiple drawers to see if we are not missing any edge case

File: packages/blade/src/components/BaseHeaderFooter/BaseHeader.tsx [hunk 6]
```diff
@@ -230,7 +234,12 @@ const _BaseHeader = ({
                 flexDirection="row"
               >
                 {title ? (
-                  <Text size="large" weight="semibold" color="surface.text.gray.normal">
+                  <Text
+                    size="large"
+                    marginTop={makeSize(size['1'])}
+                    weight="semibold"
+                    color="surface.text.gray.normal"
+                  >
```
Comment (anuraghazra) [inline packages/blade/src/components/BaseHeaderFooter/BaseHeader.tsx:242]: We are changing things on BaseHeader, make sure this doesn't not effect Modal, BottomSheet, ActionList in any negative way.

File: packages/blade/src/components/BottomSheet/BottomSheet.native.tsx [hunk 7]
```diff
@@ -77,7 +79,7 @@ const _BottomSheet = ({
     getTopOfTheStack,
   } = useBottomSheetStack();
   const currentStackIndex = getCurrentStackIndexById(id);
-  const zIndex = 100 - currentStackIndex;
+  const bottomSheetZIndex = zIndex - currentStackIndex;
```
Comment (anuraghazra) [inline packages/blade/src/components/BottomSheet/BottomSheet.native.tsx:82]: Please test things out manually once, since bottomsheet has limited tests right now.

File: packages/blade/src/components/Drawer/Drawer.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+export default {
+  title: 'Components/Drawer',
+  component: Drawer,
+  tags: ['autodocs'],
+  argTypes: {
+    ...getStyledPropsArgTypes(),
+    icon: {
+      name: 'icon',
+      type: 'select' as 'string',
```
Comment (anuraghazra) [inline packages/blade/src/components/Drawer/Drawer.stories.tsx:21]: huh?

File: packages/blade/src/components/Drawer/Drawer.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        />
+        <DrawerBody>
+          <Card padding="spacing.0">
+            <CardBody>
+              <img
+                width="100%"
+                style={{
+                  aspectRatio: '2 / 1',
+                }}
+                alt="dancing cat"
```
Comment (anuraghazra) [inline packages/blade/src/components/Drawer/Drawer.stories.tsx:55]: We love dancing cat, but he has to go 😢 let's add contextual examples.

File: packages/blade/src/components/Drawer/Drawer.web.tsx [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+  const { isMounted, isVisible } = usePresence(isOpen, {
+    enterTransitionDuration: theme.motion.duration.gentle,
+    exitTransitionDuration: theme.motion.duration.xmoderate,
+    initialEnter: true,
+  });
+
+  // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
+  const stackingLevel = drawerStack.indexOf(drawerId) + 1;
+  const isFirstDrawerInStack = stackingLevel === 1 && drawerStack.length > 1;
```
Comment (anuraghazra) [inline packages/blade/src/components/Drawer/Drawer.web.tsx:94]: Do we need to check `drawerStack.length > 1`?

File: packages/blade/src/components/Drawer/types.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+   */
+  showOverlay?: boolean;
+  /**
+   * children node.
+   *
+   * Supports DrawerHeader and DrawerBody
+   */
+  children: React.ReactNode;
+  /**
+   * zIndex property of drawer
```
Comment (anuraghazra) [inline packages/blade/src/components/Drawer/types.ts:27]: Add the `@default`

**Review by anuraghazra — DISMISSED** [review]
LGTM, minor comments.

File: packages/blade/src/components/Drawer/types.ts (hunk 11 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/types.ts:27]: Added in the documentation PR 🙈

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 10 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:94]: yes. First drawer is always stackingLevel 1 but we want this check to be true only when there is another drawer on top of it.

File: packages/blade/src/components/BaseHeaderFooter/BaseHeader.tsx (hunk 6 above)
Comment (saurabhdaware) [reply packages/blade/src/components/BaseHeaderFooter/BaseHeader.tsx:242]: nope. it is expected change for all of them. I checked on BaseHeader stories itself

File: packages/blade/src/components/BottomSheet/BottomSheet.native.tsx (hunk 7 above)
Comment (saurabhdaware) [reply packages/blade/src/components/BottomSheet/BottomSheet.native.tsx:82]: yup. Also I don't really change the value. It's vscode's variable rename thing + zIndex has same value now too

File: packages/blade/src/components/Drawer/Drawer.stories.tsx (hunk 8 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.stories.tsx:21]: Remember we discussed this with Kamlesh last time? storybook types mess. We have to do this at all places otherwise it throws error in storybook.

Anyway you can ignore it since this is temporary story I copy pasted. This prop itself won't exist in actual storybook

File: packages/blade/src/components/Drawer/Drawer.stories.tsx (hunk 9 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.stories.tsx:66]: no

Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.stories.tsx:66]: ok kidding. ignore everything regarding documentation in this PR. I have already added contextual examples in my docs PR

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 5 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.web.tsx]: changed the approach so resolving this

File: packages/blade/src/components/Drawer/Drawer.web.tsx [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+      stackingLevel: level,
+      isFirstDrawerInStack: level === 1 && drawerStack.length > 1,
+    };
+  }, [drawerId, drawerStack]);
+
+  const { refs, context } = useFloating({
+    open: isMounted,
+  });
+
+  React.useEffect(() => {
```
Comment (anuraghazra) [inline packages/blade/src/components/Drawer/Drawer.web.tsx:110]: We didn't need layoutEffect?

File: packages/blade/src/components/Drawer/Drawer.web.tsx [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+      addToDrawerStack(drawerId);
+    } else {
+      removeFromDrawerStack(drawerId);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isOpen]);
+
+  // When z-index is not defined by user, we use default drawer z index and add stackingLevel to ensure
+  // new drawer that opens, always opens on top of previous one.
+  React.useEffect(() => {
```
Comment (anuraghazra) [inline packages/blade/src/components/Drawer/Drawer.web.tsx:121]: Shouldn't this be in layout effect? Since zIndex might visually change things causing a flicker?

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 12 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:110]: nope. It's not critical for layout because anyway drawer is not visible, by the time it opens with animation the effect runs so its fine

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 13 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:121]: Nope doesn't cause flicker because this is running on isMount change which happens immediately on open and after animation on close

**Review by anuraghazra — DISMISSED** [review]

File: packages/blade/src/components/Drawer/Drawer.web.tsx [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+ *
+ * Checkout {@link https://blade.razorpay.com/?path=/docs/components-drawer Drawer Documentation}
+ * 
+ * 
+ */
+const Drawer = assignWithoutSideEffects(_Drawer, {
+  componentId: drawerComponentIds.Drawer,
+});
+
+export { Drawer };
```
Comment (snitin315) [inline packages/blade/src/components/Drawer/Drawer.web.tsx:225]: Export from `components/index.ts` as well

File: packages/blade/src/components/Drawer/DrawerContext.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (snitin315) [inline packages/blade/src/components/Drawer/DrawerContext.ts:1]: We should globally disable this rule.

File: packages/blade/src/components/Drawer/DrawerSubcomponents.native.tsx [hunk 16]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import type { DrawerHeaderProps } from './types';
+import { Text } from '~components/Typography';
+import { throwBladeError } from '~utils/logger';
+
+const DrawerHeader = (_props: DrawerHeaderProps): React.ReactElement => {
+  throwBladeError({
+    message: 'Drawer is not yet implemented for native',
+    moduleName: 'Drawer',
```
Comment (snitin315) [inline packages/blade/src/components/Drawer/DrawerSubcomponents.native.tsx:9]:
```suggestion
    message: 'DrawerHeader is not yet implemented for native',
    moduleName: 'DrawerHeader',
```

File: packages/blade/src/components/Drawer/DrawerSubcomponents.native.tsx [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+  });
+
+  return <Text>Drawer Component is not available for Native mobile apps.</Text>;
+};
+
+const DrawerBody = (_props: { children: React.ReactNode }): React.ReactElement => {
+  throwBladeError({
+    message: 'Drawer is not yet implemented for native',
+    moduleName: 'Drawer',
```
Comment (snitin315) [inline packages/blade/src/components/Drawer/DrawerSubcomponents.native.tsx:18]:
```suggestion
    message: 'DrawerBody is not yet implemented for native',
    moduleName: 'DrawerBody',
```

File: packages/blade/src/components/Drawer/Drawer.web.tsx [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+  testID,
+}: DrawerProps): React.ReactElement => {
+  const closeButtonRef = React.useRef<HTMLDivElement>(null);
+  const [zIndexState, setZIndexState] = React.useState<number>(zIndex);
+
+  useVerifyAllowedChildren({
+    children,
+    componentName: 'Drawer',
+    allowedComponents: [drawerComponentIds.DrawerHeader, drawerComponentIds.DrawerBody],
+  });
```
Comment (snitin315) [inline packages/blade/src/components/Drawer/Drawer.web.tsx:85]: Ideally this can be wrapped in __DEV__ too
```suggestion
if (__DEV__) {
  useVerifyAllowedChildren({
    children,
    componentName: 'Drawer',
    allowedComponents: [drawerComponentIds.DrawerHeader, drawerComponentIds.DrawerBody],
  });
 }
```

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 4 above)
Comment (snitin315) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:222]: We have been using MetaConstants for these too.

File: packages/blade/src/components/Drawer/types.ts [hunk 19]
```diff
@@ -0,0 +1,74 @@
+import type { TestID } from '~utils/types';
+
+type DrawerProps = {
+  /**
+   * Controlled state of drawer open or not
```
Comment (snitin315) [inline packages/blade/src/components/Drawer/types.ts:5]:
```suggestion
   * Controls the state of the drawer, indicating whether it is open or closed.
```

File: packages/blade/src/components/Drawer/types.ts [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+   */
+  isOpen: boolean;
+
+  /**
+   * Dismiss handler
```
Comment (snitin315) [inline packages/blade/src/components/Drawer/types.ts:10]:
```suggestion
   * Callback function triggered when the drawer is dismissed or closed
```

File: packages/blade/src/components/Drawer/types.ts [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+  /**
+   * zIndex property of drawer
+   */
+  zIndex?: number;
+  /**
+   *  Accessibility label for the drawer
+   */
+  accessibilityLabel?: string;
+  /**
+   *  Ref to the element to be focused on opening the drawer.
```
Comment (snitin315) [inline packages/blade/src/components/Drawer/types.ts:35]:
```suggestion
   *  Ref to the element that should receive focus when opening the drawer.
```

File: packages/blade/src/components/Drawer/types.ts [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  initialFocusRef?: React.MutableRefObject<any>;
+} & TestID;
```
Comment (snitin315) [inline packages/blade/src/components/Drawer/types.ts:39]: testID description is not present in storybook.

**Review by snitin315 — COMMENTED** [review]
For storybook docs

File: packages/blade/src/utils/componentZIndices.ts [hunk 23]
```diff
@@ -0,0 +1,10 @@
+// TODO: Move these properly to tokens at some point
+export const componentZIndices = {
```
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/utils/componentZIndices.ts:2]: Not exactly. Ideally tokens would be something like `level1`, `level2` no? still need to figure that out with design so keeping it open for now

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 14 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:225]: did it in documentation PR

File: packages/blade/src/components/Drawer/DrawerContext.ts (hunk 15 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/DrawerContext.ts:1]: Need @anuraghazra approval 🙈 

We can create issue and discuss it there

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 4 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:222]: Not really. In most components I saw we're using the name similar to the displayName of component.

MetaConstants are lower-cased values

image
image
image
image

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 18 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:85]: That breaks the react hooks rule so we'll get errors in development no? We follow similar pattern at other places

image

File: packages/blade/src/components/Drawer/types.ts (hunk 22 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Drawer/types.ts:39]: Resolved all the jsdoc comments in docs PR

File: packages/blade/src/components/Drawer/Drawer.web.tsx (hunk 18 above)
Comment (snitin315) [reply packages/blade/src/components/Drawer/Drawer.web.tsx:85]: Oh! yes, correct.
